godot-core/src/builtin/collections/extend_buffer.rs

```diff
@@ -0,0 +1,115 @@
+    buf: [MaybeUninit<T>; N],
+    len: usize,
+}
+
+impl<T, const N: usize> Default for ExtendBuffer<T, N> {
+    fn default() -> Self {
+        Self {
+            buf: [const { MaybeUninit::uninit() }; N],
+            len: 0,
+        }
+    }
+}
+
+impl<T, const N: usize> ExtendBuffer<T, N> {
+    /// Appends the given value to the buffer.
+    ///
+    /// # Panics
+    /// If the buffer is full.
+    pub fn push(&mut self, value: T) {
+        self.buf[self.len].write(value);
+        self.len += 1;
+    }
+
+    /// Returns `true` iff the buffer is full.
+    pub fn is_full(&self) -> bool {
+        self.len == N
+    }
+
+    /// Returns a slice of all initialized elements in the buffer, and sets the length of the buffer back to 0.
+    ///
+    /// It is the caller's responsibility to ensure that all elements in the returned slice get dropped!
+    pub fn drain_as_mut_slice(&mut self) -> &mut [T] {
+        // Prevent panic in self.buf[0] below.
+        let () = AssertPositive::<N>::OK;
+        debug_assert!(self.len <= N);
+        let len = self.len;
+        self.len = 0;
+        // MaybeUninit::slice_assume_init_ref could be used here instead, but it's experimental.
+        // SAFETY:
+        // - The pointer is non-null, valid and aligned.
+        // - `len` elements are always initialized.
+        // - The memory is not accessed through any other pointer, because we hold a `&mut` reference to `self`.
+        // - `len * mem::size_of::<T>()` is no larger than `isize::MAX`, otherwise the `buf` slice could not have existed either.
+        unsafe { std::slice::from_raw_parts_mut(self.buf[0].as_mut_ptr(), len) }
+    }
+}
+
+impl<T, const N: usize> Drop for ExtendBuffer<T, N> {
+    fn drop(&mut self) {
+        // Prevent panic in self.buf[0] below.
+        let () = AssertPositive::<N>::OK;
+        debug_assert!(self.len <= N);
+        // SAFETY: `slice_from_raw_parts_mut` by itself is not unsafe, but to make the resulting slice safe to use:
+        // - `self.buf[0]` is a valid pointer, exactly `self.len` elements are initialized.
+        // - The pointer is not aliased since we have an exclusive `&mut self`.
+        let slice = ptr::slice_from_raw_parts_mut(self.buf[0].as_mut_ptr(), self.len);
+        // SAFETY: the value is valid because the `slice_from_raw_parts_mut` requirements are met,
+        // and there is no other way to access the value.
+        unsafe {
+            ptr::drop_in_place(slice);
+        }
+    }
+}
+
+/// Asserts at compile time that `N` is greater than zero. The obvious solution `static_assert!(N > 0)` fails to compile:
+/// "error[E0401]: can't use generic parameters from outer item"
+/// Discussion and workaround: https://users.rust-lang.org/t/how-do-i-static-assert-a-property-of-a-generic-u32-parameter/76307/2
+struct AssertPositive<const N: usize>;
+
+impl<const N: usize> AssertPositive<N> {
+    const OK: () = assert!(N > 0, "N must be greater than zero");
```

Comment:
Could this `const`  be added directly inside `impl ExtendBuffer`, without the `AssertPositive` wrapper?

Reply:
Now gone.